src/Microsoft.Health.Fhir.Core/Features/Definition/SearchParameterDefinitionBuilder.cs

```diff
@@ -48,119 +38,90 @@ internal class SearchParameterDefinitionBuilder
+        internal static void Build(
+            IList<ITypedElement> searchParameters,
+            IDictionary<Uri, SearchParameterInfo> uriDictionary,
+            IDictionary<string, IDictionary<string, SearchParameterInfo>> resourceTypeDictionary,
+            IModelInfoProvider modelInfoProvider)
         {
+            EnsureArg.IsNotNull(searchParameters, nameof(searchParameters));
+            EnsureArg.IsNotNull(uriDictionary, nameof(uriDictionary));
+            EnsureArg.IsNotNull(resourceTypeDictionary, nameof(resourceTypeDictionary));
             EnsureArg.IsNotNull(modelInfoProvider, nameof(modelInfoProvider));
-            EnsureArg.IsNotNullOrWhiteSpace(embeddedResourceName, nameof(embeddedResourceName));
-
-            _modelInfoProvider = modelInfoProvider;
-            _assembly = assembly;
-            _embeddedResourceNamespace = embeddedResourceNamespace;
-            _embeddedResourceName = embeddedResourceName;
-        }
-
-        internal IDictionary<Uri, SearchParameterInfo> UriDictionary
-        {
-            get
-            {
-                if (!_initialized)
-                {
-                    Build();
-                }
 
-                return _uriDictionary;
-            }
-        }
-
-        internal IDictionary<string, IDictionary<string, SearchParameterInfo>> ResourceTypeDictionary
-        {
-            get
-            {
-                if (!_initialized)
-                {
-                    Build();
-                }
-
-                return _resourceTypeDictionary;
-            }
-        }
-
-        internal void Build()
-        {
-            ILookup<string, SearchParameterInfo> searchParametersLookup = ValidateAndGetFlattenedList()
-                .ToLookup(
+            ILookup<string, SearchParameterInfo> searchParametersLookup = ValidateAndGetFlattenedList(
+                searchParameters,
+                uriDictionary,
+                modelInfoProvider).ToLookup(
                     entry => entry.ResourceType,
                     entry => entry.SearchParameter);
 
             // Build the inheritance. For example, the _id search parameter is on Resource
             // and should be available to all resources that inherit Resource.
-            foreach (string resourceType in _modelInfoProvider.GetResourceTypeNames())
+            foreach (string resourceType in modelInfoProvider.GetResourceTypeNames())
             {
-                if (_resourceTypeDictionary.TryGetValue(resourceType, out IDictionary<string, SearchParameterInfo> _))
-                {
-                    // The list has already been built, move on.
-                    continue;
-                }
-
                 // Recursively build the search parameter definitions. For example,
                 // Appointment inherits from DomainResource, which inherits from Resource
                 // and therefore Appointment should include all search parameters DomainResource and Resource supports.
```

Comment:
nit: can we move this description to function summary/remarks? I find myself head scratching on recursive call inside this function, so this comment is really helpful, I just had to scroll all the way up to find it.